### ISSUE TITLE
Terraform plugins are loaded into the RAM/pod memory if they have been cached either during runtime or after the pod starts.

### ISSUE THREAD

**Issue by HEMANTHKUMAR689** [hashicorp/terraform#33526]
### Terraform Version

```shell
Terraform v1.0.9
Terraform v1.5.2
Our issue is reproducing in both of the above mentioned terraform versions.
```

### Terraform Configuration Files

case 1 : 

  provider_installation { 
            filesystem_mirror {
                   path    = "/pv/go/bin/.terraform.d/plugins"
        } 
     } 
     
case 2 : 

 plugin_cache_dir   = "/pv/go/bin/.terraform.d/plugins"

case 3 : 

  provider_installation { 
            filesystem_mirror {
                   path    = "/pv/go/bin/.terraform.d/plugins"
                   enabled = false
        } 
     }  

### Debug Output

[terraform-init.txt](https://github.com/hashicorp/terraform/files/12050433/terraform-init.txt)

### Expected Behavior

1.We are running the terraform commands inside a Kubernetes pod with Linux environment.
2.When terraform init, terraform plan, terraform apply are executed inside a Kubernetes pod, it should take the plugin from the plugin-cache
3.It should not load the plugins in to the RAM/pod memory

### Actual Behavior

1.When terraform init, terraform plan, terraform apply, terraform destroy, terraform refresh are executed inside a kubernetes pod, it is getting the plugins from the cache only.

2.It is loading the plugins in to the RAM/pod memory while performing the above terraform commands.(which is not expected)

3.After loading the plugins into the RAM/pod memory, once execution of commands was completed it is not clearing the plugins from the RAM.(The occupied RAM is not getting cleared)

Example : While provisioning the AWS terraform template, we are caching the plugins using "TF_PLUGIN_CACHE_DIR", the plugin with size of 400MB is getting downloaded in the specified cache directory(Cache directory path is inside the pod but it is mounted to an external storage using persistent volume(PV)) when terraform init is performed.In this case it is not occupying any RAM/pod memory as we have mounted the plugin-cache directory path to an external storage.But when we perform terraform plan, terraform apply, terraform destroy those plugins from the cache directory are getting loaded into the RAM/pod-memory results in the occupying of additional 400MB RAM.

4.The occupied RAM is getting cleared only after removing the plugins from the plugin-cache directory.

Note 1: We are using external storage to store the plugin-cache, they are not getting stored inside the pod.

Note 2: This is happening for the plugins that are cached during run time or after pod got started.For the plugins that are already present in the plugin-cache by the time the pod got started, those plugins are not getting loaded in to the RAM/pod memory while executing the above mentioned terraform commands. 

### Steps to Reproduce

1.Set up a Kubernetes cluster containing a pod with linux_amd64 and GO installed inside it.
2.Create a plugin-cache-dir inside a pod.
3.Create a Persistence Volume  with an external storage and mount it to a plugin-cache-dir inside the pod.
4.Create a .terraformrc file(Terraform configurations file) as mentioned in the Terraform configurations section.
5.Perform terraform init, terraform plan, terraform apply, terraform destroy to any terraform template (AWS for example)
6.Check the RAM spike inside the pod using command "docker stats | grep 'pod name' "

### Additional Context

1.We are applying the terraform commands in a CI/CD environment.
2.We are using python packages for executing those terraform commands during run time.
3.The issue is occuring even while running the above mentioned terraform commands manually inside a Kubernetes pod with linux_amd64.
4.Main thing to notice is, it is happening for the plugins that are cached during runtime or after pod start.
5.It is not happening for the plugins that are already cached by the time of pod start.
6.We have also tried keeping "TF_PLUGIN_CACHE_DIR" as environment and also using it inside the .terraformrc file and result is same in both the cases.

**Comment by apparentlymart** [comment]
Hi @HEMANTHKUMAR689! Thanks for reporting this.

Unfortunately a large part of what you've described here is from the perspective of a system you've presumably built in-house for running Terraform in Kubernetes, and since we cannot see how that system is implemented it's challenging to understand how the symptoms you've described related to behavior implemented inside Terraform itself.

Is it possible for you to describe the problem you've observed in terms of Terraform behavior rather than in terms of your external system for running Terraform? For example, it isn't clear what exactly it means for a plugin to be "occupying RAM/POD memory" but I assume what you mean is that Terraform is installing providers into a location that's inappropriate for your needs. If so, it would help to describe where you intended that Terraform would install the plugins, exactly how you've configured Terraform to achieve that, and also where Terraform is _actually_ installing the plugins that is incorrect from your perspective.

Keep in mind that we cannot see how you've configured your Kubernetes system, so we won't be able to understand what you are describing unless you can explain the situation in terms of Terraform's own behavior alone.

Thanks!

**Comment by HEMANTHKUMAR689** [comment]
@apparentlymart  Thanks for responding.
1.We are trying to run terraform in docker container.

2.We are using TF_PLUGIN_CACHE_DIR for caching the plugins

3.When we are performing terraform init those plugins are getting downloaded in our specified cache directory as expected and they are not loading in to the RAM/Memory that is assigned to the container, which is expected.

4.When I said "RAM/Memory" it means, we have assigned 3 GB of RAM for the container in which we are running terraform commands.

5.The plugins that are present in the cache directory(cache directory is pointing to an external storage outside of container) already occupied 400MB of space.

6.But when we ran terraform apply or plan or destroy instead of using plugins directly from the cache directory, the terraform is loading/copying the same plugins into the container's RAM and it is not clearing from the RAM after command execution got completed.

7.So while executing terraform plan or terraform apply or destroy, the plugins are present at two different locations (cache directory and container's RAM) at the same time which is un necessary.

7.So when we need to provision 10 different templates  of different providers or provider versions at the same time, all those 10 different plugins are getting loaded/copied in to containers RAM/Memory which is causing issues and provision failure due to insufficient RAM.

8.**MOST IMPORTANT THING HERE IS** : This is happening for the plugins that are cached during run time of container and for the plugins that already present in the cache before the start of container are not getting loaded into the container's RAM while performing any terraform commands.Terraform is directly using the plugins from the cache directory while running terraform commands.

**Comment by jbardin** [comment]
@HEMANTHKUMAR689,

It may help you diagnose what's going on by understanding what Terraform providers are under the hood. Provider plugins are just normal executable binaries, which Terraform executes and communicates with via grpc. There is nothing special about how Terraform handles the processes, so the memory management is entirely up to the operating system, and in this case, the container wrapping the processes.

**Comment by HEMANTHKUMAR689** [comment]
@jbardin  Thanks for responding.

MOST IMPORTANT THING HERE IS : This is happening for the plugins that are cached during run time of container and for the plugins that already present in the cache before the start of container are not getting loaded into the container's RAM while performing any terraform commands.Terraform is directly using the plugins from the cache directory while running terraform commands.

So once the container is deleted and restarted, the plugins that are cached are already present by the time container starts and for those plugins terraform is not copying them into the RAM of the container.It is directly using them from the cache.

**Comment by outjoker** [comment]
<img width="879" alt="image" src="https://github.com/hashicorp/terraform/assets/10362809/7e773028-2c47-4a61-aff4-e6be36584eba">
@jbardin  @apparentlymart  please check the scenario represenation over here 

@HEMANTHKUMAR689  please add if anything is needed

**Comment by jbardin** [comment]
I think it's still unclear here what _Terraform_ could change to alter the behavior you are describing, or what exactly the desired behavior would be. A plugin is a normal executable, so when terraform call `exec*` on that binary stored in the filesystem, how it's loaded into memory is up to the operating system. The memory of that plugin process is managed by the plugin itself, outside of the control of Terraform. Once the plugin process process has exited, neither Terraform not the now defunct plugin process has any control over what happens to that memory, or how it is accounted for by the container or operating system.

**Comment by justinTM** [comment]
getting some similar (i think) behavior on my end too. i specify plugin cache dir for a docker container, and mount a volume there. each terraform run (in a new container but with the same volume) the plugins are being redownloaded. 

not a big deal now that i'm on wifi, but i make many typos and often run off a hotspot... it gets old fast :)

**Comment by crw** [comment]
Hi @justinTM, thanks for the report. This may be a different issue than reported, as the previous issue had to do with memory management within containers, which is outside of the control of terraform. As such, I am going to close this issue but please feel free to open a new issue if you think you have a new problem. Note that there are other issues open that have to do with plugin cache management, so the re-downloading problem may be captured in an existing issue. Thanks!
